Add composable scenario tests for signer using madhouse

moodmosaic opened the pull request:
Adds `madhouse`-based scenario testing to signer.

Provides reusable, DRY test steps that can be composed into scenarios.

First example:
```rust
scenario![
    test_context,
    CreateFailedDkgShares,
    WriteDkgShares,
    VerifyDkgVerificationFailed,
];
```

Based on property-based state machine testing (see [spec](https://github.com/moodmosaic/stacks-core/commit/1cb033c39947d8b6c999fcf68fca3009db2e3263)).

BowTiedRadone committed:
Add madhouse-rs setup and icebreaker command

BowTiedRadone committed:
Add `CreateDkgShares` command

BowTiedRadone committed:
Add command to initialize a shared tokio runtime

BowTiedRadone committed:
Add `WriteDkgShares` command

BowTiedRadone committed:
Add `VerifyDkgVerificationFailed` command

BowTiedRadone committed:
Move `runtime` to the test context

BowTiedRadone committed:
Move `db` to the test context

BowTiedRadone committed:
Rename `CreateDkgShares` to `CreateFailedDkgShares`

BowTiedRadone committed:
Housekeep command-based testing harness

BowTiedRadone committed:
Update to latest `madhouse-rs`

moodmosaic commented:
/cc @cylewitruk

cylewitruk commented on signer/Cargo.toml:
```diff
@@ -62,6 +62,8 @@ wsts.workspace = true
 # Only for testing
 fake = { workspace = true, optional = true }
 mockall = { workspace = true, optional = true }
+madhouse = { git = "https://github.com/stacks-network/madhouse-rs.git", rev = "8623f2a4efc6a8b691f25bc95eb9801fec17b4c0" }
```
Please add this to the workspace `Cargo.toml` instead and use `madhouse.workspace = true` here.

Does this need to be a full dependency, or will it work as a `[dev-dependency]`? If it needs to be a full dependency, then please also add `optional = true` and include it in the `testing` feature.

BowTiedRadone replied:
Addressed in a1b519229b4c58bcde9d8fc82f2ebf9b4e0ed29b.

cylewitruk commented on signer/Cargo.toml:
```diff
@@ -62,6 +62,8 @@ wsts.workspace = true
 fake = { workspace = true, optional = true }
 mockall = { workspace = true, optional = true }
+madhouse = { git = "https://github.com/stacks-network/madhouse-rs.git", rev = "8623f2a4efc6a8b691f25bc95eb9801fec17b4c0" }
+proptest = { git = "https://github.com/proptest-rs/proptest.git", rev = "c9bdf18c232665b2b740c667c81866b598d06dc7" }
```
We already have `proptest` in the workspace, so `proptest.workspace = true` should be fine. This should be fine as a `[dev-dependency]` as well, no?

BowTiedRadone replied:
Addressed in a1b519229b4c58bcde9d8fc82f2ebf9b4e0ed29b.

cylewitruk reviewed (commented):
Will look more closely later, just tossing out a couple of quick comments on the dependency handling.

This repo has the convention of defining all dependencies in the workspace `Cargo.toml`, where dependencies in the `[dependencies]` stanza (i.e. full dependencies) are always specified with `default-features = false` with explicit feature enablement.

You'll also need to `cargo vet` and handle the dependency diffs.

BowTiedRadone committed:
Move `madhouse-rs` dependency to workspace `Cargo.toml`

This commit addresses the following comments:
- https://github.com/stacks-sbtc/sbtc/pull/1633#discussion_r2063117933
- https://github.com/stacks-sbtc/sbtc/pull/1633#discussion_r2063119488

BowTiedRadone committed:
Reference `madhouse-rs` dependency by tag

moodmosaic commented on signer/tests/integration/commands/dkg.rs:
```diff
@@ -0,0 +1,91 @@
+            aggregate_key,
+            dkg_shares_status: DkgSharesStatus::Failed,
+            started_at_bitcoin_block_height: 0u64.into(),
+            ..Faker.fake()
```
I think this won’t replay correctly? I’d have to check closer, or perhaps @cylewitruk knows, but I think this uses its own prng internally?

BowTiedRadone replied:
Good point. Addressed in a5df54da2a76fd8acdbdef4f1dadf9e4ec9373b8.

BowTiedRadone committed:
Make `CreateFailedDkgShares` `apply` deterministic

BowTiedRadone committed:
Rename constructor in `WriteDkgShares`

BowTiedRadone committed:
Move `madhouse-rs` to dev-dependencies

BowTiedRadone committed:
Use latest `madhouse-rs` tag

BowTiedRadone commented:
@cylewitruk: It seems no vetting is needed after moving `madhouse-rs` to dev dependencies. `cargo vet` output:
```
$ cargo vet
Vetting Succeeded (147 fully audited, 4 partially audited, 559 exempted)
 WARN Your supply-chain has unnecessary exemptions which could be relaxed or pruned.
 WARN   Consider running `cargo vet prune` to prune unnecessary exemptions and imports.
 ```

MCJOHN974 commented:
> @cylewitruk: It seems no vetting is needed after moving `madhouse-rs` to dev dependencies. `cargo vet` output:

Try to run `cargo vet prune` and it should fix problems

BowTiedRadone committed:
Run `cargo vet prune`

moodmosaic commented:
/cc @cylewitruk @djordon @aldur — are you interested in this? Should I keep open for a longer time?

djordon-stacks commented:
Closing since we haven't seen much action pushing it through, and we are unlikely to see that anytime soon. Plus, we can always reopen it later.
